File: CPP/合并排序.cpp

#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

int a[100];
void merge_sort(int *a, int start, int end);
void merge(int *a, int l, int m, int r);

int main(int argc, char const *argv[])
{
	const int len = 10;
//	int a[len] = {3,1,6,2,8,1,7,0,9,4};
	int *a = new int[len];
	cout << "original array:" << endl;
	for (int i = 0; i < len; i++)
		a[i] = rand() % 10;
	//	cout << a[i] << " ";
	cout << endl << endl;

	merge_sort(a, 0, len);

	cout << "after merge_sort" << endl;
	for (int i = 0; i < len; i++)
		cout << a[i] << " ";
	cout << endl;
	return 0;
}

void merge_sort(int *a, int start, int end)
{
/*	cout << "[";
	for (int i = start; i < end; i++)
		cout << a[i] << ",";
	cout << "\b]" << endl;*/

	if (start + 1 < end)
	{
		int mid = (start + end) / 2;
		merge_sort(a, start, mid);
		merge_sort(a, mid, end);
		merge(a, start, mid, end);
	}
	
}

void merge(int *a, int l, int m, int r)
{
	
/*	cout << "\nmerge start---------------" << endl;
	cout << "l=" << l << " m=" << m << " r=" << r << endl;
	cout << "in array" << endl;
	for (int i = l; i < r; i++)
	{
		cout << a[i] << " ";
	}
	cout << endl;*/

    int l_len = m - l;
    int r_len = r - m;

    int left[l_len];
    int right[r_len];

    int i = 0, index = 0;
    for (i = l, index = 0; i < m; i++)
        left[index++] = a[i];

    for (i = m, index = 0; i < r; i++)
        right[index++] = a[i];

    int j = 0, k = 0;
    i = l;
    while (j < l_len && k < r_len)
    {
        if (left[j] <= right[k])
            a[i++] = left[j++];
        else
            a[i++] = right[k++];
    }

    while (k < r_len)
        a[i++] = right[k++];

    while (j < l_len)
        a[i++] = left[j++];

/*    cout << "out array" << endl;
    for (int i = l; i < r; i++)
    	cout << a[i] << " ";
    cout << endl;
    cout << "merge end----------------" << endl << endl;*/
}

File: Python/合并排序.py
import random

def merge(a, b):
    '''合并两个有序的列表'''
    i, j = 0, 0
    c = []
    while i < len(a) and j < len(b):
        if a[i] < b[j]:
            c.append(a[i])
            i += 1
        else:
            c.append(b[j])
            j += 1
    c += (a[i : ] if i < len(a) else (b[j : ] if j < len(b) else None))
    return c

def merge_sort(a, l, r):
    mid = (l + r) // 2
    #首先不断分段,知道两端元素都是有序的
    if l < mid:
        merge_sort(a, l, mid)
        merge_sort(a, mid, r)
    #然后合并两个有序序列实现排序
    a[l : r] = merge(a[l : mid], a[mid: r])

a = list(random.randint(10, 50) for x in range(100))
print("original list:", a)
merge_sort(a, 0, len(a))
print("sorted  list :", a)
